share/CMakeJS.cmake

```diff
@@ -0,0 +1,640 @@
+            return()
+        endif()
+    endif()
+
+    # Needs more validation...
+    if(DEFINED ARG_NAPI_VERSION AND (ARG_NAPI_VERSION LESS_EQUAL 0))
+        message(SEND_ERROR "NAPI_VERSION for ${name} is not a valid Integer number (${ARG_NAPI_VERSION})")
+        return()
+    endif()
+
+    if(NOT DEFINED ARG_NAPI_VERSION)
+        if(NOT DEFINED NAPI_VERSION)
+            # default NAPI version to use if none specified
+            set(NAPI_VERSION 8)
+        endif()
+        set(ARG_NAPI_VERSION ${NAPI_VERSION})
+    endif()
+
+    if(ARG_ALIAS)
+        set(name_alt "${ARG_ALIAS}")
+    else()
+        set(name_alt "${ARG_NAMESPACE}")
+    endif()
+
+    if(VERBOSE)
+        message(STATUS "Configuring Napi Addon: ${name}")
+    endif()
+
+    # Begin a new Napi Addon target
+
+    add_library(${name} SHARED)
+    add_library(${name_alt}::${name} ALIAS ${name})
+
+    target_link_libraries(${name} PRIVATE cmake-js::cmake-js)
+
+    set_property(
+      TARGET ${name}
+      PROPERTY "${name}_IS_NAPI_ADDON_LIBRARY" TRUE
+    )
+
+    set_target_properties(${name}
+      PROPERTIES
+
+      LIBRARY_OUTPUT_NAME "${name}"
+      PREFIX ""
+      SUFFIX ".node"
+
+      ARCHIVE_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/lib"
+      LIBRARY_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/lib"
+      RUNTIME_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/bin"
+
+      # # Conventional C++-style debug settings might be useful to have...
+      # LIBRARY_OUTPUT_NAME_DEBUG "d${name}"
+      # ARCHIVE_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/lib/Debug"
+      # LIBRARY_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/lib/Debug"
+      # RUNTIME_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/bin/Debug"
+    )
+
+    cmakejs_napi_addon_add_sources(${name} ${ARG_UNPARSED_ARGUMENTS})
+
+    cmakejs_napi_addon_add_definitions(${name}
+      PRIVATE # These definitions only belong to this unique target
+      "CMAKEJS_ADDON_NAME=${name}"
+      "CMAKEJS_ADDON_ALIAS=${name_alt}"
+      "NAPI_CPP_CUSTOM_NAMESPACE=${ARG_NAMESPACE}"
+    )
+
+    cmakejs_napi_addon_add_definitions(${name}
+      PUBLIC # These definitions are shared with anything that links to this addon
+      "NAPI_VERSION=${ARG_NAPI_VERSION}"
+      "BUILDING_NODE_EXTENSION"
+    )
+
+endfunction()
+
+#[=============================================================================[
+Add source files to an existing Napi Addon target.
+
+cmakejs_napi_addon_add_sources(<name> [items1...])
+cmakejs_napi_addon_add_sources(<name> [BASE_DIRS <dirs>] [items1...])
+cmakejs_napi_addon_add_sources(<name> [<INTERFACE|PUBLIC|PRIVATE> [items1...] [<INTERFACE|PUBLIC|PRIVATE> [items2...] ...]])
+cmakejs_napi_addon_add_sources(<name> [<INTERFACE|PUBLIC|PRIVATE> [BASE_DIRS [<dirs>...]] [items1...]...)
+]=============================================================================]#
+function(cmakejs_napi_addon_add_sources name)
```

Comment:
What is the benefit of this over calling `target_sources` directly?

Reply:
Seperation of input validation logic.

"PRIVATE" and "PUBLIC" args don't really make sense in a ```create_addon()``` function.

It's also expandable (thinking ahead).

Reply:
> What is the benefit of this over calling `target_sources` directly?

In hindsight, it *might be* arguable that addon sources will only *ever* be PRIVATE (not accessible by any other targets). One instance where they might want to be PUBLIC might be when a user is linking two addon targets - a whole scenario I have yet to fully investigate. 

If it is determinable that PRIVATE-only sources is sufficient, the entire ```cmakejs_addon_add_sources()``` could be removed, and the call to it inside the ```create_addon()``` could be replaced with a standard ```target_add_sources(name PRIVATE file)``` type of thing.

I have no strong feelings in the initial case, do as you like. I do wonder however if such a thing might only turn into a feature request in the future - guess that depends on the scenarios like the one above...

Reply:
I have no strong feeling against having this, I was really only curious as to why it was needed.

Part of me wants to say that it seems unnecessary, but also I don't see a harm in keeping it so it may as well stay for v7